[c10d][flight recorder] remove unintended assignment of entry

shuqiangzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #119748

Summary:
auto& entry = entries_.at(*id % max_entries_);
entry = entries_.at(*id % max_entries_);
The above line of code has unintended consequence of invoking copy/assignment
of entry objects as ref itself cannot be re-assigned.

Also what could cause the crash is that the entry ref could become invalid if entries_ are
resized by other threads. and this could result in 'copy to a garbage
location'. The fix is to use a pointer which can be re-assigned after
re-acquiring the lock

Tests: python test/distributed/test_c10d_nccl.py NCCLTraceTest

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

shuqiangzhang committed:
[c10d][flight recorder] remove unintended assignment of entry
Summary:
auto& entry = entries_.at(*id % max_entries_);
entry = entries_.at(*id % max_entries_);
The above line of code has unintended consequence of invoking copy/assignment
of objects.

entries_.at(*id % max_entries_) should return a ref to the same entry
object every time. So the first entry should be valid throughout its
lifetime, we can just check if its content(id) has been overriden, and
no need to query the vector again. So the fix is simply remove the
second query and assignment

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d][flight recorder] remove unintended assignment of entry"


Summary:
auto& entry = entries_.at(*id % max_entries_);
entry = entries_.at(*id % max_entries_);
The above line of code has unintended consequence of invoking copy/assignment
of objects.

entries_.at(*id % max_entries_) should return a ref to the same entry
object every time. So the first entry should be valid throughout its
lifetime, we can just check if its content(id) has been overriden, and
no need to query the vector again. So the fix is simply remove the
second query and assignment

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d][flight recorder] remove unintended assignment of entry"


Summary:
auto& entry = entries_.at(*id % max_entries_);
entry = entries_.at(*id % max_entries_);
The above line of code has unintended consequence of invoking copy/assignment
of objects.

entries_.at(*id % max_entries_) should return a ref to the same entry
object every time. So the first entry should be valid throughout its
lifetime, we can just check if its content(id) has been overriden, and
no need to query the vector again. So the fix is simply remove the
second query and assignment

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

yifuwang commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -546,15 +546,15 @@ struct NCCLTraceBuffer {
-    auto& entry = entries_.at(*id % max_entries_);
-    if (entry.id_ == *id) {
-      update_state(entry);
+    Entry* entry = &entries_.at(*id % max_entries_);
```
Not sure if I missed anything, but it seems that the code would behave identically before and after the PR? https://godbolt.org/z/49Pn89aaG

wconstab commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -566,21 +566,21 @@ struct NCCLTraceBuffer {
       guard.lock();
 
-      // Refresh the entry ref, see if it has been overwritten
-      entry = entries_.at(*id % max_entries_);
```
@yifuwang the bug is in this line. the old code would invoke operator= and copy the requested entry contents into the value of the old reference, which could now belong to a different entry.  The new code would assign a new pointer value locally only, and let us check if it's still pointing to the same entry or not.  (it's a ring buffer)

yifuwang replied:
Ah I see it now! Thanks for highlighting

wconstab reviewed (approved):
thanks for fixing my sad bug :(

shuqiangzhang committed:
Update on "[c10d][flight recorder] remove unintended assignment of entry"


Summary:
auto& entry = entries_.at(*id % max_entries_);
entry = entries_.at(*id % max_entries_);
The above line of code has unintended consequence of invoking copy/assignment
of entry objects as ref itself cannot be re-assigned.

Also what could cause the crash is that the entry ref could become invalid if entries_ are
resized by other threads. and this could result in 'copy to a garbage
location'. The fix is to use a pointer which can be re-assigned after
re-acquiring the lock

Tests: python test/distributed/test_c10d_nccl.py NCCLTraceTest

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d][flight recorder] remove unintended assignment of entry"


Summary:
auto& entry = entries_.at(*id % max_entries_);
entry = entries_.at(*id % max_entries_);
The above line of code has unintended consequence of invoking copy/assignment
of entry objects as ref itself cannot be re-assigned.

Also what could cause the crash is that the entry ref could become invalid if entries_ are
resized by other threads. and this could result in 'copy to a garbage
location'. The fix is to use a pointer which can be re-assigned after
re-acquiring the lock

Tests: python test/distributed/test_c10d_nccl.py NCCLTraceTest

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

shuqiangzhang commented:
@pytorchbot merge

fegin commented:
@shuqiangzhang Thanks for the fix!

fegin reviewed (approved):
There seems to be a typo.

shuqiangzhang commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -528,15 +528,15 @@ struct NCCLTraceBuffer {
+        can_compute_duration = entry->time_discovered_completed_.has_value() &&
+            entry->start_ && entry->end_;
+        startEvents = entry->start_;
+        endEvents = entry->end_;
```
Merge conflicts.. Also seems pytorch trunk is broken

shuqiangzhang committed:
Update on "[c10d][flight recorder] remove unintended assignment of entry"


Summary:
auto& entry = entries_.at(*id % max_entries_);
entry = entries_.at(*id % max_entries_);
The above line of code has unintended consequence of invoking copy/assignment
of entry objects as ref itself cannot be re-assigned.

Also what could cause the crash is that the entry ref could become invalid if entries_ are
resized by other threads. and this could result in 'copy to a garbage
location'. The fix is to use a pointer which can be re-assigned after
re-acquiring the lock

Tests: python test/distributed/test_c10d_nccl.py NCCLTraceTest

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

shuqiangzhang commented:
@pytorchbot merge
